### PR TITLE
[WIP] QA Loss refactoring

### PR DESCRIPTION
**Pull request by stas00**
This is a refactoring experiment as suggested at https://github.com/huggingface/transformers/issues/6204

10 models and 1 template have been refactored - I will check for more if this looks promising (the doc and new function's signature are incomplete). Let me know whether to continue or not.

the refactoring was done with:

```
perl -0777 -pi -e '
$in = <<END;
logits = self.qa_outputs(sequence_output)
        start_logits, end_logits = logits.split(1, dim=-1)
        start_logits = start_logits.squeeze(-1)
        end_logits = end_logits.squeeze(-1)

        total_loss = None
        if start_positions is not None and end_positions is not None:
            # If we are on multi-GPU, split add a dimension
            if len(start_positions.size()) > 1:
                start_positions = start_positions.squeeze(-1)
            if len(end_positions.size()) > 1:
                end_positions = end_positions.squeeze(-1)
            # sometimes the start/end positions are outside our model inputs, we ignore these terms
            ignored_index = start_logits.size(1)
            start_positions.clamp_(0, ignored_index)
            end_positions.clamp_(0, ignored_index)

            loss_fct = CrossEntropyLoss(ignore_index=ignored_index)
            start_loss = loss_fct(start_logits, start_positions)
            end_loss = loss_fct(end_logits, end_positions)
            total_loss = (start_loss + end_loss) / 2
END
s/\Q$in\E/start_logits, end_logits, total_loss = self.calc_qa_loss(sequence_output, start_positions, end_positions)\n/msg
' \
./templates/adding_a_new_model/modeling_*    ./src/transformers/modeling_*

```

@sshleifer, I'm not sure how you're going to judge the coverage change as the coverage data is unreliable at the moment: https://github.com/huggingface/transformers/issues/6317

### CODE REVIEW

**Review by sshleifer — APPROVED** [review]
LGTM

**Comment by sgugger** [comment]
I am unsure about this as it goes against the "all in one files" policy transformers has for the models and has been highlighted as one of the main reason people like transformers in a recent survey.
Yes the alternative is duplicate code in several files, yes this is officially considered as bad computer science practice and yes it is harder for us to maintain, but the point is to have everything available in one file a researcher can easily tweak.

A similar attempt in #4944 has been ignored for the same reason.

Tagging @thomwolf and @julien-c for their thoughts.

**Comment by stas00** [comment]
I'd hold off then with completing this until if and when you give a green light to do so.

My idea of killing both rabbits with one shot would be writing an easy to maintain refactored code and have a tool that will unwrap it for those who want it unwrapped (and control the levels of how deep the unwrapped goes). Does such tool exist?

**Comment by julien-c** [comment]
> an easy to maintain refactored code and have a tool that will unfold it for those who want it unfolded (and control the levels of how deep the unfolding goes). Does such a tool exist in python land?

We explored such tools with @aaugustin a few months ago and the conclusion then was to try and build a lightweight, home-built system for this.

**Comment by sgugger** [comment]
We could add a simple script that copies the code from somewhere into the modeling files if necessary and another to check the consistency. The first could be called during `make style` and the second during `make quality`. I was thinking of doing something similar for the `TrainingArguments` and the examples this week (adding a tweakable training arguments file for each example using Trainer), so let's see how it goes for those and then continue with model refactoring the same way?

**Comment by julien-c** [comment]
(probably the same script with a different flag, like `black`, but yes, I like this idea)

**Comment by stas00** [comment]
this proved to be a failed experiment, closing this down.
